lib/widgets/new_expense.dart

```diff
@@ -0,0 +1,256 @@
+/// Creates the overlay for adding a new expense and calls the callback when the user finishes.
+class NewExpense extends StatefulWidget {
+  //Constructor
+  const NewExpense({super.key, required this.onAddExpense});
+
+  final void Function(Expense expense) onAddExpense;
+
+  @override
+  State<NewExpense> createState() {
+    return _NewExpenseState();
+  }
+}
+
+/// This is the private State class that goes with NewExpense.
+class _NewExpenseState extends State<NewExpense> {
+  final _titleController = TextEditingController(); //Controller for the title
+  final _amountController = TextEditingController(); //Controller for the amount
+  DateTime? _selectedDate; //The selected date
+  Category _selectedCategory =
+      Category.leisure; //The selected category, default is leisure
+
+  @override
+  void dispose() {
+    //Disposes the controllers when the widget is removed
+    _titleController.dispose();
+    super.dispose();
+  }
+
+  /// Opens the date picker.
+  void _presentDatePicker() async {
+    final now = DateTime.now();
+    final firstDate = DateTime(1970, 1, 1);
+    final pickedDate = await showDatePicker(
+        context: context,
+        initialDate: now,
+        firstDate: firstDate,
+        lastDate: now);
+    setState(() {
+      _selectedDate = pickedDate;
+    });
+  }
+
+  /// Shows an error dialog.
+  void _showDialog() {
+    if (Platform.isIOS) {
+      showCupertinoDialog(
+          context: context,
+          builder: (ctx) => CupertinoAlertDialog(
+                title: const Text("Invalid input"),
+                content:
+                    const Text("Please make sure you have set all values."),
+                actions: [
+                  TextButton(
+                    onPressed: () => Navigator.pop(ctx),
+                    child: const Text("Ok"),
+                  ),
+                ],
+              ));
+    } else {
+      showDialog(
+        context: context,
+        builder: (ctx) => AlertDialog(
+          title: const Text("Invalid input"),
+          content: const Text("Please make sure you have set all values."),
+          actions: [
+            TextButton(
+              onPressed: () => Navigator.pop(ctx),
+              child: const Text("Ok"),
+            ),
+          ],
+        ),
+      );
+    }
+  }
+
+  /// Saves the expense data and calls the callback.
+  void _saveExpenseData() {
+    final amount = double.tryParse(_amountController.text);
+    final invalidAmount = amount == null || amount <= 0;
+    if (_titleController.text.trim().isEmpty ||
+        invalidAmount ||
+        _selectedDate == null) {
+      _showDialog();
+      return;
+    }
+
+    widget.onAddExpense(Expense(
+        title: _titleController.text,
+        amount: amount,
+        date: _selectedDate!,
+        category: _selectedCategory));
+    Navigator.pop(context);
+  }
+
+  /// Builds the widget.
+  @override
+  Widget build(BuildContext context) {
+    var titleComponent = TextField(
```

Comment:
good you separated different parts into local variables, makes the return statement more readable